UCLID: Add support for grounding of quantifiers over finite 'groups'

ymanerka opened the pull request:
Add support for elements of type "group", and "finite_forall" and
"finite_exists" quantifiers over those groups.

polgreen commented:
Can you add some tests, and have VerifierSpec.scala run them please?

polgreen commented:
Can you check parsing for counterexamples?
~~~
    control {
        v = unroll(0);
        check;
        print_results;
        v.print_cex(myGroup);
    }
~~~

Also, how do I access elements of the group? myGroup[0]? myGroup._1?

polgreen commented:
Also some type checking issues:
~~~
    type t = integer;
    var a,b,c,d : t;
    group myGroup : t = {a,b,c,d};
    var hello : t;

    axiom test1 : finite_forall thing in myGroup :: thing < 2;   // this does not type check and it should
    axiom test2: hello > 2;  // this does
~~~

polgreen reviewed (changes requested):
More questions (+ the above issues: more tests, check the type checking):
- can I create defines that take a group as input?
- can I import a group from another module?

ymanerka commented:
Ok, I've added tests, fixed the type checking issues, and enabled the use of groups in defines. I agree that the ability to import groups from one module to another (through input variables) is useful, though I've run into issues doing so (e.g. module flattening results in the input group being referenced through a variable name other than the original name of the group). We can discuss further offline.

> 
> 
> Can you check parsing for counterexamples?
> 
> ```
>     control {
>         v = unroll(0);
>         check;
>         print_results;
>         v.print_cex(myGroup);
>     }
> ```

This currently fails because groups are expanded before we go to the solver, so the solver has no idea that groups exist. I consider this to be a good thing, because it ensures that groups don't add any extra overhead. Users can still access the relevant counterexample info by using the names of any variables they used when creating the group (e.g. a, b, c, and d in your example).

> 
> Also, how do I access elements of the group? myGroup[0]? myGroup._1?

I recall you weren't thrilled with the use of array notation for group members because we'd then have to do bounds checking to ensure we don't go outside the group (the point holds for dot notation too). The ability to access group elements by index isn't necessary for the primary use case (grounding quantifiers over the group), so it currently doesn't exist. Is there a good reason to add it?

polgreen commented:
Sorry @ymanerka, i forgot to respond to this. Could you rebase and fix the conflicts and then I'll merge as is

ymanerka commented:
Done; let me know if there are issues. Thanks!
